File: src/nl/rubensten/texifyidea/insight/LatexEnterBetweenBracesHandler.java
package nl.rubensten.texifyidea.insight;

import com.intellij.codeInsight.editorActions.enter.EnterHandlerDelegateAdapter;
import com.intellij.openapi.actionSystem.DataContext;
import com.intellij.openapi.diagnostic.Logger;
import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.editor.actionSystem.EditorActionHandler;
import com.intellij.openapi.util.Ref;
import com.intellij.psi.PsiDocumentManager;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiFile;
import com.intellij.psi.codeStyle.CodeStyleManager;
import com.intellij.util.IncorrectOperationException;
import nl.rubensten.texifyidea.file.LatexFile;
import nl.rubensten.texifyidea.psi.LatexTypes;
import org.jetbrains.annotations.NotNull;

/**
 * @author Sten Wessel
 */
public class LatexEnterBetweenBracesHandler extends EnterHandlerDelegateAdapter {

    @Override
    public Result preprocessEnter(@NotNull PsiFile file, @NotNull Editor editor, @NotNull
            Ref<Integer> caretOffset, @NotNull Ref<Integer> caretAdvance, @NotNull DataContext
            dataContext, EditorActionHandler originalHandler) {
        if (!(file instanceof LatexFile)) {
            return Result.Continue;
        }

        int offset = caretOffset.get();
        PsiElement before = file.findElementAt(offset - 1);
        PsiElement after = file.findElementAt(offset + 1);

        if (before == null || after == null) {
            return Result.Continue;
        }

        if (before.getNode().getElementType() == LatexTypes.DISPLAY_MATH_START && after.getNode().getElementType() == LatexTypes.DISPLAY_MATH_END) {
            originalHandler.execute(editor, editor.getCaretModel().getCurrentCaret(), dataContext);
            PsiDocumentManager.getInstance(file.getProject()).commitDocument(editor.getDocument());
            try {
                CodeStyleManager.getInstance(file.getProject()).adjustLineIndent(file, editor.getCaretModel().getOffset());
            }
            catch (IncorrectOperationException e) {
                Logger.getInstance(getClass()).error(e);
            }
            return Result.DefaultForceIndent;
        }

        return Result.Continue;
    }
}

File: src/nl/hannahsten/texifyidea/editor/typedhandlers/LatexEnterBetweenBracesHandler.kt
package nl.hannahsten.texifyidea.editor.typedhandlers

import com.intellij.codeInsight.editorActions.enter.EnterHandlerDelegate
import com.intellij.codeInsight.editorActions.enter.EnterHandlerDelegateAdapter
import com.intellij.openapi.actionSystem.DataContext
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.editor.Editor
import com.intellij.openapi.editor.actionSystem.EditorActionHandler
import com.intellij.openapi.util.Ref
import com.intellij.psi.PsiDocumentManager
import com.intellij.psi.PsiFile
import com.intellij.psi.codeStyle.CodeStyleManager
import com.intellij.util.IncorrectOperationException
import nl.hannahsten.texifyidea.file.LatexFile
import nl.hannahsten.texifyidea.psi.LatexTypes

/**
 * In the situation \[<caret>\], when pressing enter the caret
 * should end up with the correct indent.
 *
 * @author Sten Wessel
 */
class LatexEnterBetweenBracesHandler : EnterHandlerDelegateAdapter() {

    override fun preprocessEnter(file: PsiFile, editor: Editor, caretOffset: Ref<Int>, caretAdvance: Ref<Int>, dataContext: DataContext, originalHandler: EditorActionHandler?): EnterHandlerDelegate.Result {
        if (file !is LatexFile) {
            return EnterHandlerDelegate.Result.Continue
        }

        val offset = caretOffset.get()
        val before = file.findElementAt(offset - 1)
        val after = file.findElementAt(offset + 1)

        if (before == null || after == null) {
            return EnterHandlerDelegate.Result.Continue
        }

        if (before.node.elementType === LatexTypes.DISPLAY_MATH_START && after.node.elementType === LatexTypes.DISPLAY_MATH_END) {
            originalHandler!!.execute(editor, editor.caretModel.currentCaret, dataContext)
            PsiDocumentManager.getInstance(file.project).commitDocument(editor.document)
            try {
                CodeStyleManager.getInstance(file.project).adjustLineIndent(file, editor.caretModel.offset)
            }
            catch (e: IncorrectOperationException) {
                Logger.getInstance(javaClass).error(e)
            }

            return EnterHandlerDelegate.Result.DefaultForceIndent
        }

        return EnterHandlerDelegate.Result.Continue
    }
}
